core/shared/src/main/scala/zio/FiberPool.scala

```diff
@@ -0,0 +1,184 @@
+   *
+   * If the pool is already shutdown, this method will interrupt its caller.
+   */
+  def submit[R](task: URIO[R, Any]): URIO[R, Any] =
+    ZIO.environment[R].flatMap { r =>
+      Promise.make[Nothing, Unit].flatMap { taskReceived =>
+        val wrappedTask = task.catchAllCause(defectHandler(_).sandbox.ignore).provide(r)
+
+        state.modify {
+          case s @ State(workerCount, freeWorkers, pendingTasks, shutdownType, workers) =>
+            shutdownType match {
+              case None =>
+                if (pendingTasks.nonEmpty) {
+                  // To ensure fairness, we always queue up behind tasks that are
+                  // already pending.
+                  (
+                    taskReceived.await,
+                    State(
+                      workerCount,
+                      freeWorkers,
+                      pendingTasks.enqueue(taskReceived.succeed(()).as(wrappedTask)),
+                      shutdownType,
+                      workers
+                    )
+                  )
+
+                } else if (freeWorkers.nonEmpty) {
+                  // There were no pending tasks, but we did find an existing
+                  // free worker. So hand off the task to it.
+                  val (worker, rest) = freeWorkers.dequeue
+
+                  (worker(wrappedTask), State(workerCount, rest, pendingTasks, shutdownType, workers))
+                } else if (workerCount < workerLimit) {
+                  // No pending tasks, no free workers, but the number of workers
+                  // is below the limit. So launch a new worker.
+                  (addWorker(wrappedTask), State(workerCount + 1, freeWorkers, pendingTasks, shutdownType, workers))
+                } else {
+                  // No free resources, so queue up the task in the pending tasks
+                  // queue.
+                  (
+                    taskReceived.await,
+                    State(
+                      workerCount,
+                      freeWorkers,
+                      pendingTasks.enqueue(taskReceived.succeed(()).as(wrappedTask)),
+                      shutdownType,
+                      workers
+                    )
+                  )
+                }
+              case _ => (ZIO.interrupt, s)
+            }
+        }.flatten
+      }
+
+    }
+
+  /**
+   * Shuts down the fiber pool. Behavior is dependent on the specified [[Shutdown]].
+   */
+  def shutdown(shutdownType: Shutdown): UIO[Any] =
+    state.modify { s =>
+      shutdownType match {
+        case Shutdown.Immediate =>
+          (Fiber.interruptAll(s.workers), s.copy(shutdown = Some(shutdownType)))
+        case Shutdown.DrainRunning | Shutdown.DrainPending =>
+          (
+            ZIO.foreach(s.freeWorkers)(_.apply(ZIO.interrupt)) *> Fiber.awaitAll(s.workers),
+            s.copy(shutdown = Some(shutdownType))
+          )
+      }
+    }.flatten
```

Comment:
As before, for all of these, I'd ask: what happens if it's interrupted after the state change, but before the consequent effect is executed (or during its execution)? In many cases it may be benign, of course.